SharedMemory/RpcBuffer.cs

```diff
@@ -890,130 +894,202 @@ bool WriteProtocolV1(MessageType msgType, ulong msgId, byte[] msg, ulong respons
 
             return true;
         }
-        
-        
-        
+
+        private Boolean m_ReadThreadIsReading = false;
+        private object m_ReadThreadIsReadingLock = new object();
+
+
 
         void ReadThreadV1()
         {
-            while(true && !ReadBuffer.ShuttingDown)
+            try
             {
-                if (Interlocked.Read(ref _disposed) == 1)
-                    return;
-
-                Statistics.StartWaitRead();
+                // Work with Local Variable to prefent NPE after dispose  
+                CircularBuffer l_TempReadBuffer = ReadBuffer;
 
-                ReadBuffer.Read((ptr) =>
+                while (true && !l_TempReadBuffer.ShuttingDown)
                 {
-                    int readLength = 0;
-                    var header = FastStructure<RpcProtocolHeaderV1>.PtrToStructure(ptr);
-                    ptr = ptr + protocolLength;
-                    readLength += protocolLength;
+                    if (Interlocked.Read(ref _disposed) == 1)
```

Comment:
Should this now be: `if (Interlocked.Read(ref _disposed) != 0)` or more clearly `if (Disposed)`